### PR TITLE
[SPARK-9495] prefix of DateType/TimestampType

### PR DESCRIPTION
**Pull request by davies**
cc @rxin

### CODE REVIEW

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -160,19 +161,20 @@ public long getPrefix() {
...
-      p = PlatformDependent.UNSAFE.getLong(base, offset);
-      p = p & ((1L << numBytes * 8) - 1);
-    } else if (numBytes > 0) {
-      p = (long) PlatformDependent.UNSAFE.getInt(base, offset);
-      p = p & ((1L << numBytes * 8) - 1);
+    if (numBytes == 0) {
+      return 0;
+    }
+    long p = PlatformDependent.UNSAFE.getLong(base, offset);
+    if (ByteOrder.nativeOrder() == ByteOrder.LITTLE_ENDIAN) {
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: nativeOrder is actually a fairly expensive call. Can you move this into a static variable?

**Comment by rxin** [comment]
@davies in this case can you break the two changes into two separate pull requests?

The sort change should go in first, whereas the 2nd change should await more reviews and testing.

File: unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -160,19 +161,20 @@ public long getPrefix() {
...
-    } else  if (numBytes > 4) {
-      p = PlatformDependent.UNSAFE.getLong(base, offset);
-      p = p & ((1L << numBytes * 8) - 1);
-    } else if (numBytes > 0) {
-      p = (long) PlatformDependent.UNSAFE.getInt(base, offset);
-      p = p & ((1L << numBytes * 8) - 1);
+    if (numBytes == 0) {
+      return 0;
+    }
+    long p = PlatformDependent.UNSAFE.getLong(base, offset);
```
Comment (rxin) [inline unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: this risk segfault in the rare case of 4 byte alignment, and the remaining 4 bytes go out of the process memory boundary.

**Comment by davies** [comment]
@rxin done

**Comment by rxin** [comment]
Thanks - merging in master.

**Pull request closed without merging**